tmt/base.py

```diff
@@ -725,6 +725,8 @@ class Test(Core, tmt.export.Exportable['Test']):
 
     # Test execution data
     test: ShellScript
+    setup: List[ShellScript]
+    cleanup: List[ShellScript]
```

Comment:
Please, rebase on the latest `main`, there have been some changes that do affect your changes here. Newly it should look like this, more or less:

```python
setup: List[ShellScript] = field(
        default_factory=list,
        normalize=tmt.utils.normalize_shell_script_list)
```

And there should be no `if isinstance(..., list)` below, `normalize_shell_script_list` should take care of listification and conversion to `ShellScript` objects.

Reply:
Fixed. However, I have to also implement the following fix to be able to run the generic normalization function. Do you have any idea, where is the problem?

```diff
diff --git a/tmt/utils.py b/tmt/utils.py
index db96d470..966754ab 100644
--- a/tmt/utils.py
+++ b/tmt/utils.py
@@ -1937,9 +1937,13 @@ def dict_to_yaml(
     def _represent_path(representer: Representer, data: Path) -> Any:
         return representer.represent_scalar('tag:yaml.org,2002:str', str(data))

+    def _represent_shell_script(representer: Representer, data: ShellScript) -> Any:
+        return representer.represent_scalar('tag:yaml.org,2002:str', str(data))
+
     yaml.representer.add_representer(pathlib.Path, _represent_path)
     yaml.representer.add_representer(pathlib.PosixPath, _represent_path)
     yaml.representer.add_representer(Path, _represent_path)
+    yaml.representer.add_representer(ShellScript, _represent_shell_script)

     # Convert multiline strings
     scalarstring.walk_tree(data)
```

Reply:
I think so, yes. I'm not sure it'd work when reading tests back from `tests.yaml`.

Adding this:

```python
serialize=lambda setup: [str(script) for script in setup],
unserialize=lambda serialized: [ShellScript(script) for script in serialized]
```

should take care of both directions.

I promise I will simplify this in the future :)

Reply:
Can I use the current state of PR to reproduce the issue? If I drop the representer, what would be the command to run, let's say the included tests would crash?

I'd advise not touching `tmt.utils.SerializableContainer` for now - I do plan to include it in base classes, but it's definitely beyond the scope of this PR, way too stuff to review -, therefore I'd like to resolve it with `serialize` and `unserialize`.

Reply:
I've created a [branch](https://github.com/teemtee/tmt/tree/atomasov-setup-cleanup-test-bug) with the broken code state. This [script](https://github.com/teemtee/tmt/blob/atomasov-setup-cleanup-test-bug/tests/test/setup-cleanup/local.sh) can reproduce the bug.

Can you see the bug as well? Any ideas?

Reply:
Fixed in `Test._export()` method. Resolving thread.